File: OSX Examples/Objective-C/Convolution/Convolution/ConvolutionInstrument.m
//
//  ConvolutionInstrument.m
//  AudioKit Example
//
//  Created by Aurelius Prochazka on 6/27/12.
//  Copyright (c) 2012 Aurelius Prochazka. All rights reserved.
//

#import "ConvolutionInstrument.h"

@implementation ConvolutionInstrument

- (instancetype)init
{
    self = [super init];
    if (self) {

        // INPUTS AND CONTROLS =================================================
        _dishWellBalance = [self createPropertyWithValue:0 minimum:0 maximum:1.0];
        _dryWetBalance   = [self createPropertyWithValue:0 minimum:0 maximum:0.1];

        // INSTRUMENT DEFINITION ===============================================

        NSString *file = [AKManager pathToSoundFile:@"808loop" ofType:@"wav"];
        AKFileInput *loop = [[AKFileInput alloc] initWithFilename:file];
        [loop setLoop:YES];

        NSString *dish = [[NSBundle mainBundle] pathForResource:@"dish" ofType:@"wav"];
        NSString *well = [[NSBundle mainBundle] pathForResource:@"Stairwell" ofType:@"wav"];

        AKConvolution *dishConv;
        dishConv  = [[AKConvolution alloc] initWithInput:loop.leftOutput
                                 impulseResponseFilename:dish];


        AKConvolution *wellConv;
        wellConv  = [[AKConvolution alloc] initWithInput:loop.rightOutput
                                 impulseResponseFilename:well];


        AKMix *balance;
        balance = [[AKMix alloc] initWithInput1:dishConv
                                         input2:wellConv
                                        balance:_dishWellBalance];


        AKMix *dryWet;
        dryWet = [[AKMix alloc] initWithInput1:loop.leftOutput
                                        input2:balance
                                       balance:_dryWetBalance];

        // AUDIO OUTPUT ========================================================
        [self setAudioOutput:dryWet];

        // EXTERNAL OUTPUTS ====================================================
        // After your instrument is set up, define outputs available to others
        _auxilliaryOutput = [AKAudio globalParameter];
        [self assignOutput:_auxilliaryOutput to:dryWet];

    }
    return self;
}

@end

File: iOS Examples/Swift/Convolution/Convolution/ConvolutionInstrument.swift
//
//  ConvolutionInstrument.swift
//  Convolution
//
//  Created by Aurelius Prochazka on 2/16/15.
//  Copyright (c) 2014 Aurelius Prochazka. All rights reserved.
//

class ConvolutionInstrument: AKInstrument
{
    // INPUTS AND CONTROLS =====================================================
    
    var dishWellBalance = AKInstrumentProperty(value: 0, minimum: 0, maximum: 1.0)
    var dryWetBalance   = AKInstrumentProperty(value: 0, minimum: 0, maximum: 0.1)
    
    // INSTRUMENT DEFINITION ===================================================
    override init() {
        super.init()
        
        let file = String(NSBundle.mainBundle().pathForResource("808loop", ofType: "wav")!)
        let loop = AKFileInput(filename: file)
        
        let dish = String(NSBundle.mainBundle().pathForResource("dish",      ofType: "wav")!)
        let well = String(NSBundle.mainBundle().pathForResource("Stairwell", ofType: "wav")!)
        
        let dishConv = AKConvolution(input: loop.leftOutput,  impulseResponseFilename: dish)
        let wellConv = AKConvolution(input: loop.rightOutput, impulseResponseFilename: well)

        let balance = AKMix(input1: dishConv,        input2: wellConv, balance: dishWellBalance)
        let dryWet  = AKMix(input1: loop.leftOutput, input2: balance,  balance: dryWetBalance)
    
        setAudioOutput(dryWet)
    }
}